REF: simplify CSVFormatter

ivanovmg opened the pull request:
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Refactor CSVFormatter
----------------------
1. Put data validation in setters
2. Extract helper methods and properties

ivanovmg committed:
REF: extract properties cols and has_mi_columns

ivanovmg committed:
REF: extract property chunksize

ivanovmg committed:
REF: extract property quotechar

ivanovmg committed:
REF: extract properties data_index and nlevels

ivanovmg committed:
REF: refactor _save_chunk

ivanovmg committed:
REF: refactor _save

ivanovmg committed:
REF: extract method _save_body

ivanovmg committed:
REF: reorder _save-like methods

ivanovmg committed:
REF: extract compression property

ivanovmg committed:
REF: Extract property index_label

ivanovmg committed:
REF: extract helper properties

ivanovmg committed:
REF: delete local variables in _save_header

ivanovmg committed:
REF: extract method _get_header_rows

ivanovmg committed:
REF: move check for header into _save function

ivanovmg committed:
TYP: add several type annotations

ivanovmg committed:
FIX: fix index labels

ivanovmg committed:
FIX: fix multiindex

ivanovmg committed:
Merge branch 'master' into refactor/csvs

ivanovmg committed:
FIX: fix test failures on compression

Needed to eliminate compression setter
due to the interdependencies between ioargs and compression.

ivanovmg committed:
REF: eliminate preallocation of self.data

ivanovmg committed:
REF: extract method _convert_to_native_types

jreback reviewed (changes requested):
looks pretty good a few comments

gfyoung commented:
@ivanovmg: Welcome to `pandas` !  I like the general direction of this PR.

If you can address all of @jreback's comments, I think we should be good.

ivanovmg committed:
REF: rename regular -> flat as reviewed

ivanovmg committed:
TYP: add type annotations as reviewed

ivanovmg committed:
REF: refactor number formatting

Replace _convert_to_native_types method
in favor of a number formatting dictionary.

ivanovmg commented:
I addressed the issues raised.

* 46f593d87 (HEAD -> refactor/csvs, origin/refactor/csvs) REF: refactor number formatting
* 6041666d0 TYP: add type annotations as reviewed
* 3da7207bb REF: rename regular -> flat as reviewed

For some reason the changes are not yet reflected here in PR page.

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -127,58 +171,89 @@ def __init__(
-        # preallocate data 2d list
-        ncols = self.obj.shape[-1]
-        self.data = [None] * ncols
+    def _convert_to_native_types(self, arg, **kwargs):
```
I removed this method, but created a property ``_number_format`` (dict to store number formatting props). Now, instead of calling the method ``_convert_to_native_types`` I would call the original methods ``to_native_types`` with the ``_number_format`` passed alongside other kwargs if required.

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +76,93 @@ def __init__(
+        if isinstance(self.obj.index, ABCMultiIndex):
+            return self._get_index_label_multiindex()
+        else:
+            return self._get_index_label_regular()
```
Done.
3da7207bb REF: rename regular -> flat as reviewed

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +76,93 @@ def __init__(
+        self.chunksize = chunksize
+
+    @property
+    def index_label(self):
```
Done.
6041666d0 TYP: add type annotations as reviewed

ivanovmg replied:
ba353a5 - here I created IndexLabel type annotation in ``pandas._typing`` so that there is no code duplication in to_csv definition and CSVFormatter definition in regards to ``index_label`` type.

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +76,93 @@ def __init__(
+        return self._index_label
+
+    @index_label.setter
+    def index_label(self, index_label) -> None:
```
Done
6041666d0 TYP: add type annotations as reviewed

ivanovmg commented on pandas/io/formats/csvs.py:
```diff
@@ -222,133 +297,63 @@ def save(self) -> None:
+            for row in self._generate_multiindex_header_rows():
+                self.writer.writerow(row)
+
+    def _generate_multiindex_header_rows(self):
```
Done
6041666d0 TYP: add type annotations as reviewed

ivanovmg committed:
FIX: mypy error with index_label

ivanovmg committed:
FIX: reorder if-statements in index_label

To make sure that the newer mypy (v0.782) passes.

ivanovmg committed:
TYP: move IndexLabel to pandas._typing

This eliminates repetition of the type annotations
for index label in multiple places.

jreback commented on pandas/_typing.py:
```diff
@@ -82,6 +83,7 @@
 
 Axis = Union[str, int]
 Label = Optional[Hashable]
+IndexLabel = Optional[Union[bool, str, Sequence[Label]]]
```
shouldn't this just be `Optional[Union[Label, Sequence[Label]]]` ?

jreback replied:
cc @simonjayhawkins

ivanovmg replied:
As far as I understand, index label can be bool (at least in the original implementation). So, I left it exactly as it was.
It can probably be changed into ``Optional[Union[bool, Label, Sequence[Label]]]``, if needed.

jreback replied:
*bool* should be included in Hashable though? if that's the case then I don't think you need it here (or is somthing breaking)?

ivanovmg replied:
Replaced as IndexLabel suggested in b7dae112

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +83,92 @@ def __init__(
         self.should_close = ioargs.should_close
         self.mode = ioargs.mode
 
+        # GH21227 internal compression is not used for non-binary handles.
```
shouldn't this actually be in `get_filepath_or_buffer` instead? in `pandas/io/common.py`?

cc @twoertwein

twoertwein replied:
that is currently not in `get_filepath_or_buffer` (but it would be a good place for it!). We have the same code also in `to_csv`: https://github.com/pandas-dev/pandas/blob/bdb6e26feb26931377a92da1653ff4b95af6351d/pandas/io/formats/csvs.py#L162
If I didn't miss a function, all callers of `get_filepath_or_buffer` use the returned compression from it (if they care about compression).

ivanovmg replied:
I moved this check on compression method into ``pandas/io/common.py``. See ca888c106.

twoertwein replied:
thank you! With this change, I think we don't need the warning in `pandas/pandas/io/formats/csvs.py` line 162 anymore.

`get_filepath_or_buffer` opens some URL-like strings in binary mode (because pandas supports more operations in binary mode) even if the caller specified a non-binary mode (unless the caller explicitly specified text mode with `"t"`). If `(isinstance(filepath_or_buffer, str) and is_url(filepath_or_buffer)) or is_fsspec_url(filepath_or_buffer)`, you will need to use `fsspec_mode` instead of `mode`. I'm not sure how keen the pandas devs are on function decorates, I can imagine that it might be easier to first call `get_filepath_or_buffer` (it will decide whether to use binary/text mode) and then throw the warning.

ivanovmg replied:
@twoertwein, indeed I removed the warning from ``pandas/io/formats/csvs.py``. All of these checks are not handed to the function ``get_filepath_or_buffer``.
I am not sure I understood: is there anything else to be done on the topic? Because it seems that everything is fine now: tests pass and there is no check on the compression method in ``CSVFormatter``.

twoertwein replied:
I'm surprised that it doesn't break this test https://github.com/pandas-dev/pandas/blob/c688a0f4e39f362d1d88c35ab6093bc7963fc525/pandas/tests/io/test_gcs.py#L78 
It should call `get_filepath_or_buffer` with `mode='w'` or `mode='r'` but `get_filepath_or_buffer` will use `fsspec_mode` which adds the binary flag.

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +83,92 @@ def __init__(
+        return [""] if index_label is None else [index_label]
 
+    @property
+    def quotechar(self):
```
can you add return type annotions as much as possible on these property / functions

ivanovmg replied:
Did that.

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -75,50 +83,92 @@ def __init__(
+        return self._cols
+
+    @cols.setter
+    def cols(self, cols):
```
and type if you can

ivanovmg replied:
Added, but with the limitations.
See 1346995

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -127,58 +177,89 @@ def __init__(
-                date_format=date_format,
-                quoting=self.quoting,
-            )
+            cols = cols.to_native_types(**self._number_format)
```
great!

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -222,133 +303,63 @@ def save(self) -> None:
-            for col_loc, col in zip(b.mgr_locs, d):
-                # self.data is a preallocated list
-                self.data[col_loc] = col
+        for block in df._mgr.blocks:
```
cc @jbrockmendel this could use block.apply in the future

jbrockmendel replied:
yah i think this and one other place in the pytables code.  i'll take another swing at these today

jreback reviewed (changes requested):
thanks for the update @ivanovmg couple of questions

ivanovmg committed:
TYP: quotechar, has_mi_columns, _need_to_save...

ivanovmg committed:
TYP: chunksize, but ignored assignment check

For some reason mypy would not recognize that chunksize
turns from Optional[int] to int inside the setter.
Even setting an intentional assertion
``assert chunksize is not None``
does not help.

ivanovmg committed:
TYP: cols property

Limitations:
 - ignore type[assignment] error.
 - Created additional method _refine_cols to allow
 conversion from Optional[Sequence[Label]] to Sequence[Label].

ivanovmg committed:
TYP: nlevels and _has_aliases

ivanovmg committed:
Merge branch 'master' into refactor/csvs

ivanovmg committed:
CLN: move GH21227 check to pandas/io/common.py

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -72,46 +82,79 @@ def __init__(
+                index_label = [index_label]
+        self._index_label = index_label
+
+    def _get_index_label_from_obj(self):
```
ideally type the returns of these

ivanovmg replied:
Done

jreback reviewed (changes requested):
very nice, one small question on typing. pls ping on green.

ivanovmg committed:
TYP: remove redundant bool from IndexLabel type

ivanovmg committed:
TYP: add to _get_index_label... methods

ivanovmg committed:
TYP: use Iterator instead of Generator

ivanovmg committed:
TYP: explicitly use List type

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -120,58 +163,88 @@ def __init__(
-        # save it
-        self.cols = cols
+    @property
+    def _number_format(self) -> dict:
```
technically this should be Dict[str, Any]

jreback replied:
cc @simonjayhawkins don't we have a mypy setting that fails on this? (or should be have a code-check)?

ivanovmg replied:
Thank you! Corrected typing.

simonjayhawkins replied:
> cc @simonjayhawkins don't we have a mypy setting that fails on this? (or should be have a code-check)?

see #30539

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -120,58 +163,88 @@ def __init__(
+        return bool(self._has_aliases or self.header)
+
+    @property
+    def write_cols(self):
```
can you type here

ivanovmg replied:
Done.

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -120,58 +163,88 @@ def __init__(
-        if not index:
-            self.nlevels = 0
+    @property
+    def encoded_labels(self):
```
ca you type

ivanovmg replied:
Done.

jreback reviewed (changes requested):
sorry @ivanovmg one more set of comments, ping on green.

ivanovmg committed:
TYP: correct dict typing

ivanovmg committed:
TYP: remaining properties

ivanovmg commented:
@jreback, is there anything else to be done on this PR? Or good to go?

jreback commented:
thanks @ivanovmg very nice

simonjayhawkins commented on pandas/_typing.py:
```diff
@@ -82,6 +83,7 @@
 
 Axis = Union[str, int]
 Label = Optional[Hashable]
+IndexLabel = Optional[Union[Label, Sequence[Label]]]
```
the Optional in Label above is to include None in Label. (On hindsight, I think this could have been Union[Hashable, None] for clarity but we don't use the Union[..., None] pattern)

And also I'm not sure how we got the Ordered alias below.

In pandas._typing, I would generally prefer that we don't include the Optional, and just add it when needed in the annotations. I think this in generally allows more use of the aliases (i.e. after setting a default since we don't always set defaults in the signatures) 

so in the code, you would have 

```
index_label: Optional[IndexLabel] = None
```
instead of
```
index_label: IndexLabel = None
```

of course Label includes None anyway so the Optional is needed anyway. it's just a stylistic preference.

ivanovmg replied:
@simonjayhawkins, I agree that it is more reasonable to define Label and IndexLabel without Optional. I will take a look at this in a separate PR.

simonjayhawkins reviewed (commented):
I had just started reviewing the code as it's been merged, so this comment is just for info.
